typing/includemod.ml

```diff
@@ -1038,31 +1039,33 @@ module Functor_inclusion_diff = struct
     | Delete (Unit | Named (None,_))
     | Keep (Unit,_,_)
     | Keep (_,Unit,_)
-    | Change (_,(Unit | Named (None,_)), _) ->
+    | Change ((Unit|Named(None,_)),(Unit | Named (None,_)), _) ->
+        (* No named abstract parameters: we keep the same environment *)
         st, [||]
     | Insert (Named (Some id, arg))
     | Delete (Named (Some id, arg))
-    | Change (Unit, Named (Some id, arg), _) ->
+    | Change ((Unit|Named(None,_)), Named (Some id, arg), _)
+    | Change (Named (Some id, arg), (Unit|Named(None,_)), _) ->
+        (* One named parameter that doesn't match it anonymous counterpart,
+           we add it to the typing environment because it may
+           contain useful abbreviations. *)
         let arg' = Subst.modtype Keep st.subst arg in
         let env = Env.add_module id Mp_present arg' st.env in
         expand_params { st with env }
-    | Keep (Named (name1, _), Named (name2, arg2), _)
-    | Change (Named (name1, _), Named (name2, arg2), _) -> begin
-        let arg' = Subst.modtype Keep st.subst arg2 in
-        match name1, name2 with
-        | Some id1, Some id2 ->
-            let env = Env.add_module id1 Mp_present arg' st.env in
-            let subst = Subst.add_module id2 (Path.Pident id1) st.subst in
-            expand_params { st with env; subst }
-        | None, Some id2 ->
-            let env = Env.add_module id2 Mp_present arg' st.env in
-            { st with env }, [||]
-        | Some id1, None ->
-            let env = Env.add_module id1 Mp_present arg' st.env in
-            expand_params { st with env }
-        | None, None ->
-            st, [||]
-      end
+    | Change (Named (Some id, arg), Named(Some id2, arg2), _) ->
+        (* Change should be delete + insert: we add both abstract parameters
+           to the environment without equating them. *)
+        let arg = Subst.modtype Keep st.subst arg in
+        let env = Env.add_module id Mp_present arg st.env in
+        let arg2 = Subst.modtype Keep st.subst arg2 in
+        let env = Env.add_module id2 Mp_present arg2 env in
+        expand_params { st with env }
+    | Keep (Named (name1, _), Named (name2, arg2), _) ->
+        let arg = Subst.modtype Keep st.subst arg2 in
+        let env, subst =
+          equate_one_functor_param st.subst st.env arg name1 name2
+        in
+        expand_params { st with env; subst }
```

Comment:
The nice refactoring has the slight downside of having heterogeneous abstraction levels across the different cases here, a minor issue which could be solved by moving the Subst+Env logic to their own auxiliary functions as well.

Reply:
I have added a small auxiliary function for binding abstractions in the environment.